src/server/api/mutations/updateQuestionResponses.js

```diff
@@ -0,0 +1,115 @@
+import { log } from "../../../lib";
+import { assignmentRequiredOrAdminRole } from "../errors";
+import { cacheableData } from "../../models";
+const ActionHandlers = require("../../../integrations/action-handlers");
+
+export const updateQuestionResponses = async (
+  questionResponses,
+  campaignContactId,
+  loaders,
+  user
+) => {
+  let contact = await loaders.campaignContact.load(campaignContactId);
+  const campaign = await loaders.campaign.load(contact.campaign_id);
+  await assignmentRequiredOrAdminRole(
+    user,
+    campaign.organization_id,
+    contact.assignment_id,
+    contact
+  );
+
+  await cacheableData.questionResponse
+    .save(campaignContactId, questionResponses)
+    .catch(err => {
+      log.error(
+        `Error saving updated QuestionResponse for campaignContactID ${campaignContactId} questionResponses ${JSON.stringify(
+          questionResponses
+        )} error ${err}`
+      );
+    });
+
+  contact = await loaders.campaignContact.load(campaignContactId);
+
+  // The rest is for ACTION_HANDLERS
+  const interactionSteps = await cacheableData.campaign.dbInteractionSteps(
+    campaign.id
+  );
+
+  const stepsWithActions = interactionSteps.filter(
+    interactionStep => interactionStep.answer_actions
+  );
+
+  const actionHandlersConfigured =
+    Object.keys(ActionHandlers.rawAllActionHandlers()).length > 0;
+
+  if (stepsWithActions.length && !actionHandlersConfigured) {
+    log.error(
+      "Encountered one or more interaction steps with an action handler but no action handlers are configured"
+    );
+    return contact;
+  }
+
+  if (stepsWithActions.length && actionHandlersConfigured) {
+    const getAndProcessAction = async questionResponse => {
+      const { interactionStepId, value } = questionResponse;
+
+      const questionResponseInteractionStep = interactionSteps.find(
```

Comment:
nit: seems it would be clearer if this was done in a filter on the questionResponses to process rather than inside the promises -- none of this is async, so it seems better to have fewer promises in the first place.